cpp/src/arrow/array.cc

```diff
@@ -365,6 +365,50 @@ std::shared_ptr<Array> StructArray::field(int i) const {
   return boxed_fields_[i];
 }
 
+Status StructArray::Flatten(MemoryPool* pool, ArrayVector* out) const {
```

Comment:
This could use a few comments, particularly the `BitmapAnd` call isn't necessarily obvious.

Reply:
Will do (note this is commenting on a piece of PR #1755).